### ISSUE TITLE
Raw string format seems to not have grammar support

### ISSUE THREAD

**Issue by lichaoir** [atom/language-python#24]
When I used r'...' in my code, the colour displayed wrong as if it didn't recognise the raw string format.

**Comment by mattdeboard** [comment]
@aroben That is a good feature for "regular" strings (i.e. `u"..."` & `"..."`) but it is not appropriate to highlight `\n` in a raw byte string as a newline.

Note that these two are equivalent single-byte<sup>1</sup> strings:

```python
> ord("""
""")
10

> ord("\n")
10
```

However, this is clearly NOT the same thing:

```pytb
ord(r"\n")
TypeError: ord() expected a character, but string of length 2 found
```

We can see that it is actually a 2-byte<sup>1</sup> string:

```python
[ord(i) for i in r'\n']
> [92, 110]
```

So I think this makes it pretty clear that highlighting character sequences in raw strings which would normally be considered "escape sequences" is an error that should be corrected.

<sup>1 *They're actually more bytes than this (38 and 39 bytes, respectively), but you get my point.*</sup>

**Comment by mattdeboard** [comment]
Also, knowing nothing about the grammar mechanisms in Atom (though I'm trying to learn), what's this:

https://github.com/atom/language-python/blob/master/grammars/python.cson#L869-L870

Can this be exposed a user preference?

**Comment by aroben** [comment]
@mattdeboard You're right that `r"\n"` is a two-byte (and two-character) string. But I do think it is useful for the grammar to highlight `\n` specially in this case. In regular expressions, the two-character sequence `\n` means "a newline character", just like `\*` means "an asterisk character", etc. Since raw strings are so commonly used for defining regular expressions, the grammar highlights them as such.

**Comment by mattdeboard** [comment]
Ultimately it doesn't even matter, since the broken indentation of this
language mode makes Atom unusable for python. This renders any debate about
syntax highlighting moot.

**Comment by kbrose** [comment]
Doesn't it make more sense to have this package highlight according to the grammar of python instead of guessing what people are using raw strings for? I would vote for changing the syntax highlighting in accordance with this bug report and breaking out the regex highlighting into a separate package which can over-ride the highlighting for language-python if the user so desires.

**Comment by kbrose** [comment]
@mattdeboard Have you seen https://atom.io/packages/python-indent?

**Comment by magv** [comment]
I see that the reasoning behind this broken highlight is the belief that raw strings are only (or mainly) used for regular expressions.

I'd like to offer evidence to the contrary. Codebase I'm currently working on has lots of strings of the form `r"c:\Program Files\Some Program\Bin\Program.exe"`. As you might guess, these are not regular expressions, but Windows file paths. Currently Atom highlights `\P`, `\S` and `\B` (with different colors too); this leads to confusion, because at the first glance it looks like an indication that Python would recognize the `\` symbol as an escape, and that a double backslash `\\` is needed instead. But double backslash -- although colored by Atom like the correct thing -- is wrong, and leads to errors. Errors which are easy to overlook: never a good thing.

Please correct the hightlight by removing it from raw strings.

**Comment by lsabiao** [comment]
r''' doc string '''
breaks the grammar in the whole file...
any fix?

**Comment by VinayGupta23** [comment]
Would it be [these](https://github.com/atom/language-python/blob/c0662f3e3806dc7f093b6b4f4d554d5594c3573d/grammars/python.cson#L959-L961) lines that are causing the incorrect highlighting?

What complicates the matter further is that `\"` is special, in a double quoted raw string, when appearing at the end. In fact, it appears that raw strings have a lot of intricacies:

For example:

```python
print r"hello\"wo\rld"
# hello\"wo\rld
```
But this gives:

```python
print r"Hello\"
# SyntaxError 
```
And this:

```python
print r"Hello\\", r"Hello\""
# Hello\\ Hello\"
```

So how does one make a raw string with a single backslash at the end?

**P.S. Funnily, even Github's syntax highlighter gets it wrong!**
